### PR TITLE
BUG: stats: Fix `zipf.pmf` and `zipfian.pmf` for int32 `k`

### LINKED ISSUE #20692: BUG: stats.zipf: incorrect pmf values
### Describe your issue.

The pmf value of zipf may return an incorrect value.

If the shape parameter 'a' is set to 9, pmf returns an incorrect value.

Scipy returns the following for inputs greater than or equal to 0:
0.000000000000000000e+00
9.979956327307618613e-01
1.949210220177269260e-03
5.070343101817618650e-05
3.807051211283729024e-06
5.109737639581500849e-07
9.903013870737536426e-08
2.473126213304208103e-08
7.435646897038533250e-09
2.576001169444504624e-09
9.979956327307618846e-10
**0.000000000000000000e+00**
1.154001579655571043e-09
4.953901915407116765e-10
0.000000000000000000e+00
0.000000000000000000e+00

In wolfram, the values will be as follows:
PDF[zipfdistribution[8], 11]
1/(2357947691 ζ(9))≈4.23248×10^-10

(Note: that in wolfram's definition, the shape parameter has a difference of 1.)

In my experiment, I obtained the following.
0
0.9979956327307621568646761321051
0.001949210220177269837626320570518
5.070343101817620062311010171748e-5
3.807051211283730151613907364292e-6
5.109737639581502243147141796378e-7
9.903013870737539184201191741694e-8
2.473126213304208857623746328563e-8
7.435646897038535452370912820884e-9
2.576001169444505442417827654193e-9
9.979956327307621568646761321051e-10
4.232475709872573915655519654635e-10
1.93418239662842562191429526205e-10
9.411058434986077082628878685352e-11

The cause of this bug is that there is no overflow avoidance when the shape paramter and input values to pmf are both integer types.

### Reproducing Code Example

```python
import numpy as np
import scipy.stats as stats

x = np.arange(0, 16).astype(np.int32)
dist = stats.zipf(9)
pmf = dist.pmf(x)
```

### Error message

```shell
(N/A)
```

### SciPy/NumPy/Python version and system information

```shell
1.13.0 1.26.4 sys.version_info(major=3, minor=9, micro=5, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /c/opt/64/include
    lib directory: /c/opt/64/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= ZEN MAX_THREADS=24
    pc file directory: c:/opt/64/lib/pkgconfig
    version: 0.3.26.dev
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /c/opt/64/include
    lib directory: /c/opt/64/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= ZEN MAX_THREADS=24
    pc file directory: c:/opt/64/lib/pkgconfig
    version: 0.3.26.dev
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.12.0
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.10
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  pythran:
    include directory: C:\Users\runneradmin\AppData\Local\Temp\pip-build-env-z8a5b3_f\overlay\Lib\site-packages/pythran
    version: 0.15.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
Python Information:
  path: C:\Users\runneradmin\AppData\Local\Temp\cibw-run-0mr_3znw\cp39-win_amd64\build\venv\Scripts\python.exe
  version: '3.9'
```

### PR DESCRIPTION
**Pull request by dschmitz89**
#### Reference issue
Closes gh-20692

In the long run, these issues should hopefully be taken care of by the infrastructure (#15928 ).

### CODE REVIEW

**Comment by dschmitz89** [comment]
`zipfian` likely suffers from the same issue: https://github.com/scipy/scipy/blob/2f166227c4be835586e3887a0ba201fb537e6fa5/scipy/stats/_discrete_distns.py#L1415

I am not familiar with these distributions though and could not cook up a failing example in a quick session, so let me know if this would be important to fix as well.

File: scipy/stats/tests/test_discrete_distns.py [hunk 1]
```diff
@@ -627,3 +627,14 @@ def test_betanbinom_kurtosis(self, n, a, b, ref):
...
+
+class TestZipf:
+    def test_gh20692(self):
+        # test that int32 data for k generates same output as double
+        k = np.arange(0, 1000)
+        k_int32 = k.astype(np.int32)
+        dist = zipf(9)
+        pmf = dist.pmf(k)
+        pmf_k_int32 = dist.pmf(k_int32)
+        assert_equal(pmf, pmf_k_int32)
```
Comment (tylerjereddy) [inline scipy/stats/tests/test_discrete_distns.py:640]: Hey Daniel, I just messed around for a bit and found this case that failed before/passed after a similar patch per your comment on `zipfian`:

<details>

```diff
diff --git a/scipy/stats/_discrete_distns.py b/scipy/stats/_discrete_distns.py
index cda7c4e11..2ee610827 100644
--- a/scipy/stats/_discrete_distns.py
+++ b/scipy/stats/_discrete_distns.py
@@ -1412,7 +1412,8 @@ class zipfian_gen(rv_discrete):
         return 1, n
 
     def _pmf(self, k, a, n):
-        return 1.0 / _gen_harmonic(n, a) / k**a
+        k = k.astype(np.float64)
+        return 1.0 / _gen_harmonic(n, a) * k**-a
 
     def _cdf(self, k, a, n):
         return _gen_harmonic(k, a) / _gen_harmonic(n, a)
diff --git a/scipy/stats/tests/test_discrete_distns.py b/scipy/stats/tests/test_discrete_distns.py
index c5993ebde..bbe9085ce 100644
--- a/scipy/stats/tests/test_discrete_distns.py
+++ b/scipy/stats/tests/test_discrete_distns.py
@@ -627,3 +627,12 @@ class TestBetaNBinom:
         #      return float(fourth_moment/var**2 - 3.)
         assert_allclose(betanbinom.stats(n, a, b, moments="k"),
                         ref, rtol=3e-15)
+
+
+def test_zipfian_gh_20692():
+    k = np.arange(0, 1000)
+    k_int32 = k.astype(np.int32)
+    dist = zipfian(111, 22)
+    pmf = dist.pmf(k)
+    pmf_k_int32 = dist.pmf(k_int32)
+    assert_equal(pmf, pmf_k_int32)
```

</details>

You shouldn't trust my judgement on `stats` stuff, but it does look quite similar in nature as you note. The obtained `pmf` floats are ridiculously tiny, maybe `111` is a bit much, but anyway I just pushed it to error out for integers. It does run lightning fast at least.

Comment (dschmitz89) [reply scipy/stats/tests/test_discrete_distns.py:648]: Added in the last commit.

**Comment by dschmitz89** [comment]
I went ahead and merged this since it is a quite simple bug fix.

**Pull request merged**